File: Data_structures/Binary Search Tree/BST.java
public class BST{
	int key;
	BST left,right;
	private static final BST NIL=new BST();
	private BST(){
		left=right=this;
	}
	public BST(int k){
		key=k;
		left=right=NIL;
	}
	public BST(int k,BST l,BST r){
		key=k;
		left=l;
		right=r;
	}
	public BST insert(int k){
		return insert(this,k);
	}
	private BST insert(BST b,int k){
		if(b==NIL)
			return new BST(key);
		if(k<b.key)
			return left=left.insert(k);
		if(k>b.key)
			return right=right.insert(k);
		return this;
	}

	public String toString(){
		if(this==NIL)
				return "";
		String buf="";
		if(left!=NIL)
			buf+=left.toString()+",";
		buf+=key;
		if(right!=NIL)
			buf+=","+right.toString();
		return buf;
	}

	public static void main(String[] args) {
		BST tree=new BST(45);
		tree.insert(20);
		tree.insert(25);
		tree.insert(47);
		tree.insert(22);
		System.out.println(tree);
	}


}

File: Data_structures/Binary Search Tree/cpp/bst.cpp
#include <iostream>
#include <queue>

using std::cout;
using std::endl;
using std::queue;

class TreeNode {
    public:
        int val_;
        TreeNode* left = nullptr;
        TreeNode* right = nullptr;

        TreeNode(int val) {
            val_ = val;
        }
};

// Insert a new node and return the root of the tree.
TreeNode* insert(TreeNode* root, int val) {
    if (!root) {
        return new TreeNode(val);
    }

    if (val > root->val_) { 
        root->right = insert(root->right, val);
    } else if (val < root->val_) {
        root->left = insert(root->left, val);
    }
    return root;
}

// Return the minimum value node of the BST.
TreeNode* minValueNode(TreeNode* root) {
    TreeNode* curr = root;
    while (curr && curr->left) {
        curr = curr->left;
    }
    return curr;
}

// Remove a node and return the root of the tree.
TreeNode* remove(TreeNode* root, int val) {
    if (!root) {
        return new TreeNode(val);
    }

    if (val > root->val_) { 
        root->right = remove(root->right, val);
    } else if (val < root->val_) {
        root->left = remove(root->left, val);
    } else {
        if (!root->left) {
            return root->right;
        } else if (!root->right) {
            return root->left;
        } else {
            TreeNode* minNode = minValueNode(root->right);
            root->val_ = minNode->val_;
            root->right = remove(root->right, minNode->val_);
        }
    }
    return root;
}

bool search(TreeNode* root, int target) {
    if (!root) {
        return false;
    }

    if (target > root->val_) {
        return search(root->right, target);
    } else if (target < root->val_) {
        return search(root->left, target);
    } else {
        return true;
    }
}